test/framework/src/main/java/org/elasticsearch/cluster/coordination/LinearizabilityChecker.java

```diff
@@ -224,21 +225,41 @@ public static boolean isLinearizable(SequentialSpec spec, History history, Funct
      * @param spec the sequential specification of the datatype
      * @param history the history of events to check for linearizability
      * @param missingResponseGenerator used to complete the history with missing responses
-     * @param terminateEarly a condition upon which to terminate early
      * @return true iff the history is linearizable w.r.t. the given spec
      */
-    public static boolean isLinearizable(
-        SequentialSpec spec,
-        History history,
-        Function<Object, Object> missingResponseGenerator,
-        BooleanSupplier terminateEarly
-    ) {
-        history = history.clone(); // clone history before completing it
-        history.complete(missingResponseGenerator); // complete history
-        final Collection<List<Event>> partitions = spec.partition(history.copyEvents());
-        return partitions.stream().allMatch(h -> isLinearizable(spec, h, terminateEarly));
+    public static boolean isLinearizable(SequentialSpec spec, History history, Function<Object, Object> missingResponseGenerator)
+        throws LinearizableCheckAborted {
+        final ScheduledThreadPoolExecutor scheduler = Scheduler.initScheduler(Settings.EMPTY, "test-scheduler");
+        final AtomicBoolean abort = new AtomicBoolean();
+        try {
+            history = history.clone(); // clone history before completing it
+            history.complete(missingResponseGenerator); // complete history
+            final Collection<List<Event>> partitions = spec.partition(history.copyEvents());
+            // Large histories can be problematic and have the linearizability checker run OOM
+            // Bound the time how long the checker can run on such histories (Values empirically determined)
+            if (history.size() > 300 || partitions.stream().anyMatch(p -> p.size() > 25)) {
+                logger.warn("Detected large history or partition for linearizable check. Limiting execution time");
+                scheduler.schedule(() -> abort.set(true), 10, TimeUnit.SECONDS);
+            }
+            var allLinearizable = partitions.stream().allMatch(h -> isLinearizable(spec, h, abort::get));
+            if (abort.get()) {
```

Comment:
Maybe we should do just what we did in https://github.com/elastic/elasticsearch/pull/101357/files#diff-c1fe1b667c241e62ccc27b117ca2daf92b358596f8756e2e9c7a8eec2a8ba4e9L446-L448 and return true in that case? that will avoid having to introduce the new exception and catching it in the calling site.

Reply:
I think there is a value on distinguishing between `is linearizable` and `could not perform the check`. I am imagining some change could break something and will silently result in `could not perform the check` for all runs. Exception might not be the most elegant solution, we might use enum instead.

Reply:
Makes sense and since this is a checked exception, callers should handle it anyway 👍